source/blog/2015-09-22-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,218 @@
+title: How the RSpec 3.0 composable matchers work
+author: Sam Phippen
+---
+
+##Warmup: the `===` protocol
+
+In Ruby, a number of objects implement the `===` protocol. Some of the objects
+that implement this protocol include:
+
+* all class objects
+* lambdas
+* regexes
+* ranges
+
+Conceptually, this protocol represents a "match", but can be thought of as
+categorising instances of objects. That is, the object that implements the
+`===` declares whether or not it belongs in the same category as the other
+object.  For example, `Array === []` returns `true`. This is because an empty
+array is an instance of the array class. `Array === {}` returns false because
+a hash is not an array. For regexes, `===` returns true if the passed string
+matches the regex, and false if it does not.
+
+The `===` method is used in Ruby case statements to decide if a branch should
+be taken. For example
+
+```Ruby
+case 3
+when (1...18)
+  puts "hello"
+end
+```
+
+will print hello, because `(1...18) === 3` returns `true`.
+
+The new RSpec matchers make heavy use of `===` internally, if you'd like to get
+more of an understanding of `===`, I'd heavily suggest you pop open IRB and
+play around with it a little before moving on.
+
+## RSpec matchers
+
+In an RSpec expectation expression like `expect(object_1).to eq(object_2)`
+there are two primary objects at play, the **target** and the **matcher**. The
+**target** object is created by "expect" and exposes `to` (and `to_not`) which
+takes the matcher is is responsible for invoking `matches?` on it.
+
+It's possible to see this happening in an IRB session, e.g:
+
+```Ruby
+>> require "rspec/expectations"
+=> true
+>> extend(RSpec::Matchers)
+=> main
+>> expect(3).to Object.new
+NoMethodError: undefined method `matches?' for #<Object:0x007fa20d1f4e68>
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:50:in `block in handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
+    from /Users/sam/.gem/ruby/2.1.5/gems/rspec-expectations-3.2.1/lib/rspec/expectations/expectation_target.rb:54:in `to'
+    from (irb):4
+    from /Users/sam/.rubies/ruby-2.1.5/bin/irb:11:in `<main>'
+>>
+```
+
+In this session, we extend the RSpec DSL into the main object and then call
+`expect(3).to Object.new`. This blows up because our new object does not
+implement the `matches?` method.
+
+In RSpec expectations matchers implement `matches?` to provide assertions for
+your test suite, some of these are complex but they can also be quite simple.
+For example, here's the logic of `matches?` for the `eq` matcher:
+
+```ruby
+def matches?(actual)
+  actual == expected
+end
+```
+
+## Composable RSpec matchers
+
+A composable RSpec matcher is one which expects other RSpec matchers for it's
```

Comment:
I think this sentence would be more accurate if you say "accepts other RSpec matchers" instead of "expects other RSpec matchers". "Expects" suggests that you can't pass non-matchers as arguments.